File: app/src/main/java/com/elementary/tasks/core/view_models/google_tasks/GoogleTaskListsViewModel.java
package com.elementary.tasks.core.view_models.google_tasks;

import android.app.Application;

import com.elementary.tasks.core.app_widgets.UpdatesHelper;
import com.elementary.tasks.core.cloud.Google;
import com.elementary.tasks.core.data.models.GoogleTask;
import com.elementary.tasks.core.data.models.GoogleTaskList;
import com.elementary.tasks.core.utils.SuperUtil;
import com.elementary.tasks.core.view_models.Commands;
import com.google.api.services.tasks.model.Task;
import com.google.api.services.tasks.model.TaskList;
import com.google.api.services.tasks.model.TaskLists;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

import androidx.lifecycle.LiveData;

/**
 * Copyright 2018 Nazar Suhovich
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
public class GoogleTaskListsViewModel extends BaseTaskListsViewModel {

    public LiveData<List<GoogleTaskList>> googleTaskLists;

    public GoogleTaskListsViewModel(Application application) {
        super(application);
        googleTaskLists = getAppDb().googleTaskListsDao().loadAll();
    }

    public void sync() {
        Google mGoogle = Google.getInstance(getApplication());
        if (mGoogle == null || mGoogle.getTasks() == null) {
            return;
        }
        boolean isConnected = SuperUtil.isConnected(getApplication());
        if (!isConnected) {
            result.postValue(Commands.FAILED);
            return;
        }
        isInProgress.postValue(true);
        run(() -> {
            TaskLists lists = null;
            try {
                lists = mGoogle.getTasks().getTaskLists();
            } catch (IOException e) {
                e.printStackTrace();
            }
            if (lists != null && lists.size() > 0 && lists.getItems() != null) {
                for (TaskList item : lists.getItems()) {
                    String listId = item.getId();
                    GoogleTaskList taskList = getAppDb().googleTaskListsDao().getById(listId);
                    if (taskList != null) {
                        taskList.update(item);
                    } else {
                        Random r = new Random();
                        int color = r.nextInt(15);
                        taskList = new GoogleTaskList(item, color);
                    }
                    getAppDb().googleTaskListsDao().insert(taskList);
                    List<Task> tasks = mGoogle.getTasks().getTasks(listId);
                    if (tasks.isEmpty()) {
                        end(() -> {
                            isInProgress.postValue(false);
                            result.postValue(Commands.UPDATED);
                            UpdatesHelper.getInstance(getApplication()).updateTasksWidget();
                        });
                    } else {
                        List<GoogleTask> googleTasks = new ArrayList<>();
                        for (Task task : tasks) {
                            GoogleTask googleTask = getAppDb().googleTasksDao().getById(task.getId());
                            if (googleTask != null) {
                                googleTask.setListId(listId);
                                googleTask.update(task);
                            } else {
                                googleTask = new GoogleTask(task, listId);
                            }
                            googleTasks.add(googleTask);
                        }
                        getAppDb().googleTasksDao().insertAll(googleTasks);
                        end(() -> {
                            isInProgress.postValue(false);
                            result.postValue(Commands.UPDATED);
                            UpdatesHelper.getInstance(getApplication()).updateTasksWidget();
                        });
                    }
                }
            }
        });
    }

    public void reload() {

    }

    public void clearList(GoogleTaskList googleTaskList) {
        Google mGoogle = Google.getInstance(getApplication());
        if (mGoogle == null || mGoogle.getTasks() == null) {
            return;
        }
        boolean isConnected = SuperUtil.isConnected(getApplication());
        if (!isConnected) {
            result.postValue(Commands.FAILED);
        } else {
            isInProgress.postValue(true);
            run(() -> {
                List<GoogleTask> googleTasks = getAppDb().googleTasksDao().getAllByList(googleTaskList.getListId(), Google.TASKS_COMPLETE);
                getAppDb().googleTasksDao().deleteAll(googleTasks);
                mGoogle.getTasks().clearTaskList(googleTaskList.getListId());
                end(() -> {
                    isInProgress.postValue(false);
                    result.postValue(Commands.UPDATED);
                    UpdatesHelper.getInstance(getApplication()).updateTasksWidget();
                });
            });
        }
    }
}

File: app/src/main/java/com/elementary/tasks/core/view_models/google_tasks/GoogleTaskListsViewModel.kt
package com.elementary.tasks.core.view_models.google_tasks

import android.content.Context
import com.elementary.tasks.core.app_widgets.UpdatesHelper
import com.elementary.tasks.core.cloud.GTasks
import com.elementary.tasks.core.data.dao.GoogleTaskListsDao
import com.elementary.tasks.core.data.dao.GoogleTasksDao
import com.elementary.tasks.core.data.models.GoogleTask
import com.elementary.tasks.core.data.models.GoogleTaskList
import com.elementary.tasks.core.utils.Prefs
import com.elementary.tasks.core.utils.WorkManagerProvider
import com.elementary.tasks.core.utils.launchDefault
import com.elementary.tasks.core.utils.withUIContext
import com.elementary.tasks.core.view_models.Commands
import com.elementary.tasks.core.view_models.DispatcherProvider
import com.google.api.services.tasks.model.TaskLists
import kotlinx.coroutines.Job
import timber.log.Timber
import java.io.IOException
import java.util.Random

class GoogleTaskListsViewModel(
  prefs: Prefs,
  context: Context,
  gTasks: GTasks,
  dispatcherProvider: DispatcherProvider,
  workManagerProvider: WorkManagerProvider,
  updatesHelper: UpdatesHelper,
  googleTasksDao: GoogleTasksDao,
  googleTaskListsDao: GoogleTaskListsDao
) : BaseTaskListsViewModel(
  prefs,
  context,
  gTasks,
  dispatcherProvider,
  workManagerProvider,
  updatesHelper,
  googleTasksDao,
  googleTaskListsDao
) {

  val googleTaskLists = googleTaskListsDao.loadAll()
  val allGoogleTasks = googleTasksDao.loadAll()
  val defTaskList = googleTaskListsDao.loadDefault()
  private var isSyncing = false
  private var job: Job? = null

  init {
    defTaskList.observeForever { }
  }

  fun loadGoogleTasks() {
    if (!gTasks.isLogged) {
      postCommand(Commands.FAILED)
      return
    }
    postInProgress(true)
    job = launchDefault {
      var lists: TaskLists? = null
      try {
        lists = gTasks.taskLists()
      } catch (e: IOException) {
        e.printStackTrace()
      }

      if (lists != null && lists.size > 0 && lists.items != null) {
        for (item in lists.items) {
          val listId = item.id
          var taskList = googleTaskListsDao.getById(listId)
          if (taskList != null) {
            taskList.update(item)
          } else {
            val r = Random()
            val color = r.nextInt(15)
            taskList = GoogleTaskList(item, color)
          }
          Timber.d("loadGoogleTasks: $taskList")
          googleTaskListsDao.insert(taskList)
          val tasksList = gTasks.getTasks(listId)
          if (tasksList.isNotEmpty()) {
            for (task in tasksList) {
              var googleTask = googleTasksDao.getById(task.id)
              if (googleTask != null) {
                googleTask.update(task)
                googleTask.listId = task.id
              } else {
                googleTask = GoogleTask(task, listId)
              }
              googleTasksDao.insert(googleTask)
            }
          }
        }
        val local = googleTaskListsDao.all()
        if (local.isNotEmpty()) {
          val listItem = local[0].apply {
            this.def = 1
            this.systemDefault = 1
          }
          googleTaskListsDao.insert(listItem)
        }
      }

      withUIContext {
        postInProgress(false)
      }
      job = null
    }
  }

  fun sync() {
    if (!gTasks.isLogged) {
      postCommand(Commands.FAILED)
      return
    }
    if (isSyncing) return
    isSyncing = true
    postInProgress(true)
    launchDefault {
      var lists: TaskLists? = null
      try {
        lists = gTasks.taskLists()
      } catch (e: IOException) {
        e.printStackTrace()
      }

      if (lists != null && lists.size > 0 && lists.items != null) {
        for (item in lists.items) {
          val listId = item.id
          var taskList = googleTaskListsDao.getById(listId)
          if (taskList != null) {
            taskList.update(item)
          } else {
            val r = Random()
            val color = r.nextInt(15)
            taskList = GoogleTaskList(item, color)
          }
          googleTaskListsDao.insert(taskList)
          val tasks = gTasks.getTasks(listId)
          if (tasks.isEmpty()) {
            withUIContext {
              postInProgress(false)
              postCommand(Commands.UPDATED)
              updatesHelper.updateTasksWidget()
            }
          } else {
            val googleTasks = ArrayList<GoogleTask>()
            for (task in tasks) {
              var googleTask = googleTasksDao.getById(task.id)
              if (googleTask != null) {
                googleTask.listId = listId
                googleTask.update(task)
              } else {
                googleTask = GoogleTask(task, listId)
              }
              googleTasks.add(googleTask)
            }
            googleTasksDao.insertAll(googleTasks)
            withUIContext {
              postInProgress(false)
              postCommand(Commands.UPDATED)
              updatesHelper.updateTasksWidget()
            }
          }
        }
      }
      isSyncing = false
    }
  }
}
